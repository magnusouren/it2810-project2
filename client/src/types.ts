export interface User {
  name: string;
  id: string;
  loginState: boolean;
  ratings?: Rating[];
  watchlist?: Movie[];
}

/**
 * The Movie type is based on the API response from
 * https://developers.themoviedb.org/3/movies/get-popular-movies
 * The API response is a bit more complex, but we only need these fields.
 */
export interface Movie {
  adult: boolean;
  backdrop_path: string;
  genre_ids: number[];
  _id: number;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string;
  release_date: string;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}

/**
 * The Category type is based on the API response from
 * https://developers.themoviedb.org/3/genres/get-movie-list
 */
export interface Category {
  id: number;
  name: string;
}

export interface Rating {
  movieId: number;
  rating: number;
}

export type AlphabeticalSort = 'a-z' | 'z-a' | '';

export type RatingSort = 'h-l' | 'l-h' | '';
