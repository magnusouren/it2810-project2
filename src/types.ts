export type User = {
  name: string;
  movieList?: Movie[];
};

export type Movie = {
  id: number;
  title: string;
  year: number;
  rating: number;
  poster: string;
};
