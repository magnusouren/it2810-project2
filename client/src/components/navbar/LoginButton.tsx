import { AccountCircle as AccountCircleIcon } from '@mui/icons-material';
import MenuIcon from '@mui/icons-material/Menu';
import { Box, IconButton, MenuItem } from '@mui/material';
import Popover from '@mui/material/Popover';
import { useState } from 'react';
import { Link } from 'react-router-dom';

import { useUser } from '../../context/UserContext';
import DarkModeToggle from '../darkModeToggle/DarkModeToggle';
import styles from './LoginButton.module.scss';

/**
 * Login button component.
 * If the user is logged in, the user's name is displayed.
 * If the user is not logged in, the login button is displayed.
 * If the user is logged in, a menu is displayed with the following options:
 * - My Watchlist
 * - Logout
 * - Delete user
 *
 * @returns {React.JSX.Element}
 */

const LoginButton = () => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const { user, login, logout, deleteUser } = useUser();

  const handleLogin = () => {
    login();
    handleMenuClose();
  };

  const handleLogout = () => {
    logout();
    handleMenuClose();
  };

  const handleDeleteUser = () => {
    deleteUser();
  };

  const handleMenuOpen = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
    if (event.key === 'Tab') {
      if (event.shiftKey) {
        return;
      } else {
        handleMenuClose();
      }
    }
  };

  return user ? (
    <>
      <IconButton color='inherit' onClick={handleMenuOpen} data-testid='menu'>
        <p className={styles.loginText} data-testid='user-name'>
          {user.name}
        </p>
        <MenuIcon fontSize='large' aria-label='menu' titleAccess='menu' sx={{ margin: '0 0 0 10px' }} />
      </IconButton>
      <Popover
        open={Boolean(anchorEl)}
        anchorEl={anchorEl}
        onClose={handleMenuClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
      >
        <MenuItem>
          <Box sx={{ padding: '0 10px' }}>
            <DarkModeToggle />
          </Box>
        </MenuItem>
        <Link to='/watchlist' className={styles.navLink} data-testid='watchlist-link' onClick={handleMenuClose}>
          <MenuItem>My Watchlist</MenuItem>
        </Link>
        <button className={styles.button} onClick={handleLogout} data-testid='logout'>
          <MenuItem>Logout</MenuItem>
        </button>
        <button
          className={styles.button}
          onKeyDown={handleKeyDown}
          onClick={handleDeleteUser}
          data-testid='delete-user'
        >
          <MenuItem>Delete user</MenuItem>
        </button>
      </Popover>
    </>
  ) : (
    <IconButton color='inherit' onClick={handleLogin} data-testid='login-button'>
      <p className={styles.loginText}>Login</p>
      <AccountCircleIcon fontSize='large' sx={{ margin: '0 0 0 6px' }} />
    </IconButton>
  );
};

export default LoginButton;
