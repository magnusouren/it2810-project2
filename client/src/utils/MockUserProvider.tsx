import { FC, ReactNode, useState } from 'react';
import { vi } from 'vitest';

// Importer UserContext
import { UserContext } from '../context/UserContext'; // Importer UserContext
import { User } from '../types';

interface MockUserProviderProps {
  children: ReactNode;
  mockUser?: User;
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
}

const MockUserProvider: FC<MockUserProviderProps> = ({
  children,
  mockUser,
  loginMock = vi.fn(),
  logoutMock = vi.fn(),
  deleteUserMock = vi.fn(),
  toggleDarkModeMock = vi.fn(),
}) => {
  const [user, setUser] = useState<User | undefined>(mockUser);

  const login = loginMock;
  const logout = logoutMock;
  const deleteUser = deleteUserMock;
  const toggleDarkMode = toggleDarkModeMock;

  const value = {
    user,
    setUser,
    login,
    logout,
    deleteUser,
    darkMode: false,
    toggleDarkMode,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default MockUserProvider;
