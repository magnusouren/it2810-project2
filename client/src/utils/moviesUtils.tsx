import { GET_MOVIES, GET_MOVIES_BY_GENRE, GET_MOVIES_BY_RATING, GET_MOVIES_BY_TITLE_AZ } from '../graphql/queries';

export const determineQueryAndVariables = (page: number, genre: string, sort: string, userId: string) => {
  const baseVariables = { page: page, userId: userId };
  const alphabeticalSort = sort === 'a-z' || sort === 'z-a';
  const ratingSort = sort === 'h-l' || sort === 'l-h';

  if (genre) {
    if (alphabeticalSort) {
      return {
        query: GET_MOVIES_BY_TITLE_AZ,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    if (ratingSort) {
      return {
        query: GET_MOVIES_BY_RATING,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    return {
      query: GET_MOVIES_BY_GENRE,
      variables: { ...baseVariables, genreId: genre },
    };
  }

  if (alphabeticalSort) {
    return {
      query: GET_MOVIES_BY_TITLE_AZ,
      variables: { ...baseVariables, order: sort },
    };
  }
  if (ratingSort) {
    return {
      query: GET_MOVIES_BY_RATING,
      variables: { ...baseVariables, order: sort },
    };
  }
  return {
    query: GET_MOVIES,
    variables: { ...baseVariables },
  };
};
