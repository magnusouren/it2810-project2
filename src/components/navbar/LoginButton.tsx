import { AccountCircle as AccountCircleIcon, PlaylistPlay as PlaylistPlayIcon } from '@mui/icons-material';
import { IconButton, Menu, MenuItem } from '@mui/material';
import { useState } from 'react';
import { Link } from 'react-router-dom';

import { useUser } from '../../context/UserContext';
import styles from './LoginButton.module.scss';

/**
 * Login button component.
 * If the user is logged in, the user's name is displayed.
 * If the user is not logged in, the login button is displayed.
 * If the user is logged in, a menu is displayed with the following options:
 * - My Watchlist
 * - Logout
 * - Delete user
 *
 * @returns {JSX.Element}
 */
function LoginButton() {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const userContext = useUser();
  if (!userContext) throw new Error('UserContext is null');
  const { user, login, logout, deleteUser } = userContext;

  const handleLogin = () => {
    login();
    handleMenuClose();
  };

  const handleLogout = () => {
    logout();
    handleMenuClose();
  };

  const handleDeleteUser = () => {
    deleteUser();
  };

  const handleMenuOpen = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
  };

  return user ? (
    <>
      <IconButton color='inherit' onClick={handleMenuOpen} data-testid='menu'>
        <p className={styles.loginText} data-testid='user-name'>
          {user.name}
        </p>
        <PlaylistPlayIcon />
      </IconButton>
      <Menu anchorEl={anchorEl} open={Boolean(anchorEl)} onClose={handleMenuClose}>
        <Link to='/watchlist' className={styles.navLink} data-testid='watchlist-link'>
          <MenuItem onClick={handleMenuClose}>My Watchlist</MenuItem>
        </Link>
        <MenuItem onClick={handleLogout} data-testid='logout'>
          Logout
        </MenuItem>
        <MenuItem onClick={handleDeleteUser} data-testid='delete-user'>
          Delete user
        </MenuItem>
      </Menu>
    </>
  ) : (
    <IconButton color='inherit' onClick={handleLogin} data-testid='login-button'>
      <AccountCircleIcon />
      <p className={styles.loginText}>Login</p>
    </IconButton>
  );
}

export default LoginButton;
