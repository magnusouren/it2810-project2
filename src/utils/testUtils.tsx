import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { UserProvider } from '../context/UserContext';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouterAndUserContext = (component: JSX.Element) =>
  render(
    <UserProvider>
      <BrowserRouter>{component}</BrowserRouter>
    </UserProvider>,
  );

const queryClient = new QueryClient();

// Function to render a component wrapped inside a browserrouter and queryclient for testing purposes
export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );
