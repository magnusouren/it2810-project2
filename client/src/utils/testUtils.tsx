import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { User } from '../types';
import MockUserProvider from './MockUserProvider';

const link = new HttpLink({
  uri: 'http://localhost:4000', // Endepunktet til din GraphQL server
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

type ProviderMocks = {
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
};

interface ProviderProps {
  child: React.ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mocks?: any[];
  mockUser?: User;
  providerMocks?: ProviderMocks;
}

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithProviders = ({ child, mockUser, mocks, providerMocks }: ProviderProps) =>
  render(
    <MockUserProvider mockUser={mockUser} {...providerMocks}>
      <ApolloProvider client={client}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <BrowserRouter>{child}</BrowserRouter>
        </MockedProvider>
      </ApolloProvider>
    </MockUserProvider>,
  );
