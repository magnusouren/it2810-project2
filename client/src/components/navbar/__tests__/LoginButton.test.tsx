import { cleanup, fireEvent, screen } from '@testing-library/react';
import { vi } from 'vitest';

import { User } from '../../../types';
import { renderWithProviders } from '../../../utils/testUtils';
import LoginButton from '../LoginButton';

const mockUser: User = {
  id: '1',
  name: 'Foo Bar',
  loginState: true,
};

describe('LoginButton', () => {
  beforeEach(() => {
    cleanup();
    window.localStorage.clear();
  });

  it('Should match snapshot when logged inn', () => {
    const { container } = renderWithProviders({ child: <LoginButton />, mockUser: mockUser });
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot when logged out', () => {
    const { container } = renderWithProviders({ child: <LoginButton /> });
    expect(container).toMatchSnapshot();
  });

  it('Should render a login button when the user is not logged in', () => {
    renderWithProviders({ child: <LoginButton /> });
    expect(screen.getByText('Login')).toMatchSnapshot();
  });

  it('Should render the user name when the user is logged in', async () => {
    renderWithProviders({ child: <LoginButton />, mockUser: mockUser });
    expect(screen.getByText(mockUser.name)).toBeDefined();
  });

  it('Should render the menu when it is clicked', async () => {
    renderWithProviders({ child: <LoginButton />, mockUser: mockUser });
    fireEvent.click(screen.getByTestId('menu'));
    expect(screen.getByText('My Watchlist')).toBeDefined();
  });

  it('Should call login when clicking login', async () => {
    const login = vi.fn();
    renderWithProviders({ child: <LoginButton />, providerMocks: { loginMock: login } });
    expect(login).not.toHaveBeenCalled();
    fireEvent.click(screen.getByText('Login'));
    expect(login).toHaveBeenCalled();
  });

  it('Should move to watchlist page', async () => {
    renderWithProviders({ child: <LoginButton />, mockUser: mockUser });

    fireEvent.click(screen.getByTestId('menu'));
    fireEvent.click(screen.getByText('My Watchlist'));

    expect(window.location.pathname).toBe('/watchlist');
  });

  it('Should log out when clicking logout', async () => {
    const logout = vi.fn();
    renderWithProviders({
      child: <LoginButton />,
      mockUser: mockUser,
      providerMocks: { logoutMock: logout },
    });
    expect(logout).not.toHaveBeenCalled();
    fireEvent.click(screen.getByTestId('menu'));
    fireEvent.click(screen.getByTestId('logout'));
    expect(logout).toHaveBeenCalled();
  });

  it('Should log out when clicking delete user', async () => {
    const delteUser = vi.fn();
    renderWithProviders({
      child: <LoginButton />,
      mockUser: mockUser,
      providerMocks: { deleteUserMock: delteUser },
    });
    expect(delteUser).not.toHaveBeenCalled();
    fireEvent.click(screen.getByTestId('menu'));
    fireEvent.click(screen.getByTestId('delete-user'));
    expect(delteUser).toHaveBeenCalled();
  });
});
