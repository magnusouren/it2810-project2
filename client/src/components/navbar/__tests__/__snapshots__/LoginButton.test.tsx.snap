// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginButton > Should match snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-testid="login-button"
    tabindex="0"
    type="button"
  >
    <p
      class="_loginText_67d6a9"
    >
      Login
    </p>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-jj92jx-MuiSvgIcon-root"
      data-testid="AccountCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`LoginButton > Should render a login button when the user is not logged in 1`] = `
<p
  class="_loginText_67d6a9"
>
  Login
</p>
`;
